scripts/rest_service.py

```diff
@@ -0,0 +1,300 @@
+
+            self._log.info("accept ws for %s" % cid)
+            try:
+                # keep this async task alive as long as there are messages to
+                # send and the websocket is alive
+                while True:
+                    msg = client.get_msg()
+                    if msg:
+                        await ws.send_json(msg)
+                    else:
+                        await asyncio.sleep(0.1)
+
+            except WebSocketDisconnect:
+                self._log.info("dropped ws for %s" % cid)
+        # ----------------------------------------------------------------------
+
+    def _status_callback(self, cid: str, job: psij.Job, status: psij.JobStatus) -> None:
+
+        client = self._clients.get(cid)
+        if not client:
+            print("unknown client cid %s" % cid)
+            return
+
+        msg = {'jobid': job.id,
+               'time': status.time,
+               'message': status.message,
+               'state': str(status.state),
+               'metadata': status.metadata,
+               'exit_code': status.exit_code}
+        self._log.debug('cb: %s: %s', cid, msg)
+        client.send(msg)
+
+    def _request_register(self, name: str, url: Optional[str] = None) -> str:
+        '''
+        parameters:
+            name:str : name of psij executor to use
+            url:str  : optional url to pass to psij executor
+
+        returns:
+            uid:str  : unique ID identifying the registered client
+            addr:str : URL to be used for state notifications
+        '''
+
+        # register new client
+        cid = 'client.%04d' % self._cnt
+        self._cnt += 1
+
+        # create executor
+        jex = psij.JobExecutor.get_instance(name=name, url=url)
+
+        # register state callback for this cid
+        cb = functools.partial(self._status_callback, cid)
+        jex.set_job_status_callback(cb)
+
+        # store client information
+        self._clients[cid] = _Client(jex)
+
+        # client is now known and initialized
+        return cid
+
+    def _request_submit(self, cid: str, spec: Dict[str, Any]) -> str:
+        '''
+        parameters:
+           cid:str   : client ID
+           spec:Dict : job spec
+
+        returns:
+           jobid:str : psij job ID for submitted job
+        '''
+
+        client = self._clients.get(cid)
+        if not client:
+            raise ValueError('unknown client cid %s' % cid)
+
+        job = psij.Job(self._deserialize.from_dict(spec, 'JobSpec'))
+        client.add_job(job)
+        client.jex.submit(job)
+
+        return job.id
+
+    def _request_cancel(self, cid: str, jobid: str) -> None:
+        '''
+        parameters:
+           cid:str   : client ID
+           jobid:str : psij job ID for job to be canceled
+        '''
+
+        client = self._clients.get(cid)
+        if not client:
+            raise ValueError('unknown client cid %s' % cid)
```

Comment:
Maybe just call ``self._clients[cid]`` and let a KeyError be thrown? Not important though

Reply:
I prefer the explicit error message in this case.

Reply:
Seems reasonable